File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    # (현재 인덱스, 현재까지의 합)
    queue = deque([(0, 0)])
    answer = 0
    
    while queue:
        index, current_sum = queue.popleft()
        
        # 모든 숫자를 사용한 경우
        if index == len(numbers):
            if current_sum == target:
                answer += 1
            continue
        
        # 현재 숫자를 더하거나 빼는 경우
        queue.append((index + 1, current_sum + numbers[index]))
        queue.append((index + 1, current_sum - numbers[index]))
    
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    private int answer = 0;  // 클래스 멤버 변수
    
    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0, 0);
        return answer;
    }
    
    private void dfs(int[] numbers, int target, int index, int currentSum) {
        // 모든 숫자를 다 사용한 경우
        if (index == numbers.length) {
            if (currentSum == target) {
                answer++;
            }
            return;
        }
        
        // 현재 숫자를 더하는 경우
        dfs(numbers, target, index + 1, currentSum + numbers[index]);
        
        // 현재 숫자를 빼는 경우
        dfs(numbers, target, index + 1, currentSum - numbers[index]);
    }
}